symtab: optimize, add test and fix linking issue

ycsin opened the pull request:
- Added a function to sanitize function names in the Python script
- Renamed the `start_addr` to `first_addr`
- Test symtab - somehow I forgot to add tests for it
- Added a new `Debug: Symtab` area in the maintainer list, LMK if this should just go into `Debug`
- Use linkonce (similar to `sw_isr_table`) to fix linking issue with external toolchain

[Doc preview](https://builds.zephyrproject.io/zephyr/pr/73333/docs/services/debugging/index.html)

ycsin committed:
build: symtab: sanitize function name

Sanitize the function name and only store the valid part.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
debug: symtab: rename `start_addr` to `first_addr`

`start_addr` is the address of the first symbol, rename it to
`first_addr` instead as it seems more intuitive and relatable
to the comments.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: debug: symtab: add test

Add test for symtab.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
MAINTAINERS: add Symtab

Add symtab to the list.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt reviewed (dismissed):
LGTM -

If unsure, common iterator bounds are `begin` and `end` (taken from C++), where `end` usually lies outside of the range so that `for (it = begin; it < end; ..)` works.

One thing I was curious about was whether the symbol table is sorted numerically by address. If it's sorted, then you can do a binary search on it based on the address, which is O(log(n)) (i.e. pretty good).

I haven't looked into too much detail to see how it's done though.

ycsin committed:
build: symtab: Add a dummy entry at the end for the binary search

The `symtab_find_symbol_name()` is using an adapted binary
search function to get the entry between 2 addresses, we need
to add a dummy entry at the end so that the search function
can remain simple and straightforward without doing
out-of-bound checks:

   20  \
       |
       |
   50  x
       |
       |
   90  x
    .  |
    .  |
    .  |
dummy  /

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
doc: services: debugging: mipi: fix heading

Fix heading level.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

nashif commented on doc/services/debugging/mipi_stp_decoder.rst:
```diff
@@ -17,7 +17,7 @@ written to Stimulus Port registers maps directly to STP stream.
 This module can be used to perform on-chip decoding of the data stream. STP v2 is used.
 
 Usage
-#####
```
that is unrelated change, isnt it?

ycsin replied:
Not really related to the original intent of this PR, but something seems off while I'm here:

![Screenshot 2024-05-27 at 10 32 46 PM](https://github.com/zephyrproject-rtos/zephyr/assets/1563974/e572e70f-f2ab-4713-9a0e-45ce17da6565)

I guess the `Usage` and `Limitations` probable should stay inside the MIPI documentation

ycsin replied:
![Screenshot 2024-05-28 at 11 07 41 AM](https://github.com/zephyrproject-rtos/zephyr/assets/1563974/d4fb3fab-0154-4ebb-94ef-af21d430d1c9)

What it would look like after this PR

nashif replied:
looks good

ycsin committed:
doc: services: debugging: add documentation for symtab

Add a simple documentation for symbol table.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
build: symtab: prevent entries with the same address

Append new entry to the symtab list only if it has unique
address.

Added a bit more comments and move the debug print to after
the list is sorted.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
debug: symtab: fix linking issue due to variable redefinition

Employs the same linkonce magic of sw_isr_table to fix the
multiple definition of the symtab variables issue that I
get in my application build that doesn't use `west`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
